### PR TITLE
Hyperbridge Adapter

### PR DESCRIPTION
**Pull request by seunlanlege**
Adds the `HyperbridgeReporter` and `HyperbridgeAdapter`. There are a few decision to be made here:

- [x] How should payments be handled? stablecoins or native tokens?
- [x] How should messages be relayed? Self-relay or pay for 3rd-party relayers ?

Once these questions are answered, this code should be ready to merge

### CODE REVIEW

**Comment by geogons** [comment]
1. I would say native tokens as that's how we handle it with other adapters currently

2. We could do self relay as we are running other relayer infra for other adapters as well. What's the relayer repo so we can take a look?

**Comment by seunlanlege** [comment]
> I would say native tokens as that's how we handle it with other adapters currently

ok so this makes things a bit complicated, Hyperbridge collects fees ultimately in stablecoins, so we need to provide sufficient native tokens that can be swapped for the stablecoins using some local uniswapV2.

https://github.com/gnosis/hashi/blob/7e8179f1b65865bcb58b822b7f2009d041b44ee7/packages/evm/contracts/adapters/Reporter.sol#L54-L59

 So either this function takes an extra field for fees which is calculated offchain, Or we ask the uniswapV2 router for a quote which can be vulnerable to flashloan attacks.
 
 > What's the relayer repo so we can take a look?
 
 Relayer repo is [here](https://github.com/polytope-labs/hyperbridge/tree/main/tesseract), relevant docs are [here](https://docs.hyperbridge.network/developers/network/relayer)

**Comment by allemanfredi** [comment]
Hey @seunlanlege ! Thanks for the PR. Can't you simply approve and dispatch? The user (or someone else) would transfer some USDC (or another stablecoin), and then, when `_dispatch` is invoked, you approve the fee and call `IDispatcher(host()).dispatch(post)`. Another solution that would consume less gas is to approve an infinite amount within the constructor, so you don't have to approve the amount every time that `_dispatch` is invoked.

**Comment by seunlanlege** [comment]
> Hey @seunlanlege ! Thanks for the PR. Can't you simply approve and dispatch? The user (or someone else) would transfer some USDC (or another stablecoin), and then, when `_dispatch` is invoked, you approve the fee and call `IDispatcher(host()).dispatch(post)`. Another solution that would consume less gas is to approve an infinite amount within the constructor, so you don't have to approve the amount every time that `_dispatch` is invoked.

Yes this is the strategy I've implemented as well. Just wanted to be sure this approach would be fine.

**Comment by allemanfredi** [comment]
> > Hey @seunlanlege ! Thanks for the PR. Can't you simply approve and dispatch? The user (or someone else) would transfer some USDC (or another stablecoin), and then, when `_dispatch` is invoked, you approve the fee and call `IDispatcher(host()).dispatch(post)`. Another solution that would consume less gas is to approve an infinite amount within the constructor, so you don't have to approve the amount every time that `_dispatch` is invoked.
> 
> Yes this is the strategy I've implemented as well. Just wanted to be sure this approach would be fine.

Ah perfect! i didn't notice the approve within the constructor. Regarding the fee i think it could be a constant defined by you or a variable that can be changes by an owner

**Comment by seunlanlege** [comment]
> Regarding the fee i think it could be a constant defined by you or a variable that can be changes by an owner

Unfortunately Hyperbridge fees are dynamic based on the size of the message.

https://github.com/polytope-labs/hyperbridge/blob/main/evm/src/hosts/EvmHost.sol#L1082

If you pay in stablecoins, then there's no extra overhead. Paying in native tokens will require: 

1. Estimate offchain the required native tokens, which can be swapped for dispatch fees
2. Set the new fee on the reporter contract
3. Dispatch the message (although price for the native asset in the pool could've changed by this point)

Alternatively, if we proceed with the stablecoin approach, then you only need to periodically fund the reporter contract in stablecoins. Everything else would just work.

**Comment by allemanfredi** [comment]
> > Regarding the fee i think it could be a constant defined by you or a variable that can be changes by an owner
> 
> Unfortunately Hyperbridge fees are dynamic based on the size of the message.
> 
> https://github.com/polytope-labs/hyperbridge/blob/main/evm/src/hosts/EvmHost.sol#L1082
> 
> If you pay in stablecoins, then there's no extra overhead. Paying in native tokens will require:
> 
> 1. Estimate offchain the required native tokens, which can be swapped for dispatch fees
> 2. Set the new fee on the reporter contract
> 3. Dispatch the message (although price for the native asset in the pool could've changed by this point)
> 
> Alternatively, if we proceed with the stablecoin approach, then you only need to periodically fund the reporter contract in stablecoins. Everything else would just work.

I would go with the stablecoin approach as well because it's less gas expensive.

**Comment by seunlanlege** [comment]
> I would go with the stablecoin approach as well because it's less gas expensive.

My thoughts as well, in that case, this is good to be merged

**Comment by allemanfredi** [comment]
> > I would go with the stablecoin approach as well because it's less gas expensive.
> 
> My thoughts as well, in that case, this is good to be merged

I need to test it before merging it.

**Comment by seunlanlege** [comment]
> My thoughts as well, in that case, this is good to be merged

No problem, these are the supported test networks: https://docs.hyperbridge.network/developers/evm/contract-addresses

File: packages/evm/contracts/adapters/Hyperbridge/HyperbridgeAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    event ReporterSet(uint256 indexed chainId, uint16 indexed chainId, address indexed reporter);
+
+    function setReporterByChain(uint256 chainId, uint16 chainId, address reporter) external onlyOwner {
+        bytes32 stateMachineId = keccak256(StateMachine.evm(chainId));
+        enabledReportersPaths[stateMachineId] = keccak256(reporter);
+        chainIds[stateMachineId] = chainId;
+        emit ReporterSet(chainId, chainId, reporter);
+    }
+
+    /// Process incoming blockhashes
```
Comment (allemanfredi) [inline packages/evm/contracts/adapters/Hyperbridge/HyperbridgeAdapter.sol]: remove this comment please as it process block hashes but also message hashes

File: packages/evm/contracts/adapters/Hyperbridge/HyperbridgeAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+contract HyperbridgeAdapter is BlockHashAdapter, Ownable, BaseIsmpModule {
+    mapping(uint32 => bytes32) public enabledReporters;
+    mapping(uint32 => uint256) public chainIds;
+
+    error UnauthorizedRequest();
+
+    event ReporterSet(uint256 indexed chainId, uint16 indexed chainId, address indexed reporter);
+
+    function setReporterByChain(uint256 chainId, uint16 chainId, address reporter) external onlyOwner {
```
Comment (allemanfredi) [inline packages/evm/contracts/adapters/Hyperbridge/HyperbridgeAdapter.sol]: why 2 `chainId`?

File: packages/evm/contracts/adapters/Hyperbridge/HyperbridgeReporter.sol [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            // recipient chain
+            dest: StateMachine.evm(targetChainId),
+            // recipient contract
+            to: abi.encode(adapter),
+            // serialized message
+            body: payload,
+            // no timeouts
+            timeout: 0,
+            // requests will be self-relayed
+            fee: 0,
```
Comment (allemanfredi) [inline packages/evm/contracts/adapters/Hyperbridge/HyperbridgeReporter.sol]: just to be sure. Is it normal 0 here?

**Review by allemanfredi — CHANGES_REQUESTED** [review]
Can you add also the deployment tasks please?

File: packages/evm/contracts/adapters/Hyperbridge/HyperbridgeReporter.sol (hunk 3 above)
Comment (seunlanlege) [reply packages/evm/contracts/adapters/Hyperbridge/HyperbridgeReporter.sol]: Yes this fee is for 3rd party relayers, @geogons has noted they won't be neccessary

File: packages/evm/contracts/adapters/Hyperbridge/HyperbridgeAdapter.sol (hunk 2 above)
Comment (seunlanlege) [reply packages/evm/contracts/adapters/Hyperbridge/HyperbridgeAdapter.sol]: oversight on my part

**Comment by seunlanlege** [comment]
> Can you add also the deployment tasks please?

I'm unfortunately unfamiliar with the hardhat framework 😅 .

**Comment by allemanfredi** [comment]
> > Can you add also the deployment tasks please?
> 
> I'm unfortunately unfamiliar with the hardhat framework 😅 .

you can copy one!

File: packages/evm/contracts/adapters/Hyperbridge/HyperbridgeReporter.sol [hunk 4]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.20;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { Reporter } from "../Reporter.sol";
+
```
Comment (allemanfredi) [inline packages/evm/contracts/adapters/Hyperbridge/HyperbridgeReporter.sol]: rm this new line.

**Review by allemanfredi — CHANGES_REQUESTED** [review]
How did you test the deployment without importing `hyperbridge.ts` within `tasks/deploy/adapters/index.ts`?

**Review by allemanfredi — APPROVED** [review]
LGTM!

**Pull request merged**